File: csharp/src/EasyCollaboration.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using Ch.Elca.Iiop.Idl;
using log4net;
using omg.org.CORBA;
using tecgraf.openbus.core.v2_0.services.access_control;
using tecgraf.openbus.core.v2_0.services.offer_registry;
using tecgraf.openbus.services.collaboration.v1_0;
using IComponent = scs.core.IComponent;
using System.Threading;

namespace tecgraf.openbus.services.collaboration.easy
{
  public class EasyCollaboration : IEasyCollaboration
  {
    #region Fields

    private static readonly ILog Logger =
      LogManager.GetLogger(typeof(EasyCollaboration));

    private OpenBusContext _context;
    private SessionRegistry _sessions;
    private CollaborationRegistry _collabs;
    private CollaborationSession _theSession;

    private CollaborationObserver _observer;
    private EventConsumer _consumer;
    private int _subsId;
    private int _obsId;

    private readonly object _locker = new object();

    #endregion

    /// <summary>
    /// Construtor padrão, nesse caso a biblioteca vai usar implementações 
    /// básicas de <see cref="EventConsumer"/> e <see cref="CollaborationObserver"/>.
    /// <seealso cref="EasyCollaboration.Consumer"/>
    /// </summary>
    /// <param name="context">contexto do OpenBus, objeto responsável pelo gerenciamento de conexões</param>
    public EasyCollaboration(OpenBusContext context)
    {
      _context = context;
      _observer = new EasyCollaboration.Observer();
      _consumer = new EasyCollaboration.Consumer();
    }

    /// <summary>
    /// Construtor opcional para aplicação fornecer um objeto para observação da sessão.
    /// <para>Esse construtor é opcional porque a aplicação sempre pode adicionar outros consumidores ou outros observadores no <see cref="CollaborationSession"/> quando inicia a colaboração através do <see cref="EasyCollaboration.StartCollaboration()"/>.</para>
    /// </summary>
    /// <param name="context">contexto do OpenBus, objeto responsável pelo gerenciamento de conexões</param>
    /// <param name="observer">observador da sessão de colaboração</param>
    /// <exception cref="ArgumentException">caso o <c>observer</c> não implemente <see cref="MarshalByRefObject"/> além da interface <see cref="CollaborationObserver"/></exception>
    public EasyCollaboration(OpenBusContext context, CollaborationObserver observer)
      : this(context)
    {
      if ((observer != null) && ! (observer is MarshalByRefObject))
        throw new ArgumentException("CollaborationObserver object must implement MarshalByRefObject also", "observer");
      _observer = observer;
    }

    /// <summary>
    /// Construtor opcional para aplicação fornecer tanto o consumidor quanto o observador da sessão.
    /// <para>Esse construtor é opcional porque a aplicação sempre pode adicionar outros consumidores ou outros observadores no <see cref="CollaborationSession"/> quando inicia a colaboração através do <see cref="EasyCollaboration.StartCollaboration()"/>.</para>
    /// </summary>
    /// <param name="context">contexto do OpenBus, objeto responsável pelo gerenciamento de conexões</param>
    /// <param name="consumer">consumidor do canal de eventos</param>
    /// <param name="observer">observador da sessão de colaboração</param>
    /// <exception cref="ArgumentException">caso o <c>consumer</c> não implemente <see cref="MarshalByRefObject"/> além da interface <see cref="EventConsumer"/></exception>
    /// <exception cref="ArgumentException">caso o <c>observer</c> não implemente <see cref="MarshalByRefObject"/> além da interface <see cref="CollaborationObserver"/></exception>
    public EasyCollaboration(OpenBusContext context, EventConsumer consumer, CollaborationObserver observer)
      : this(context, observer)
    {
      if ((consumer != null) && ! (consumer is MarshalByRefObject))
        throw new ArgumentException("EventConsumer object must implement MarshalByRefObject also", "consumer");
      _consumer = consumer;
    }

    public CollaborationSession StartCollaboration()
    {
      Logger.Info("Starting collaboration");

      lock (_locker)
      {
        SessionRegistry sreg = GetSessions();
        try
        {
          _theSession = sreg.getSession();
          Logger.Info("Session retrieved: " + OrbServices.GetSingleton().object_to_string(_theSession));
        }
        catch (TargetInvocationException e)
        {
          if (e.InnerException is SessionDoesNotExist)
            Logger.Warn("Session not found for entity " + (e.InnerException as SessionDoesNotExist).entity);
        }
        catch (Exception e)
        {
          Logger.Error("Error trying to obtain session: " + e);
          throw;
        }

        if (_theSession == null)
        {
          try
          {
            CollaborationRegistry collab = GetCollabs();
            _theSession = collab.createCollaborationSession();
            sreg.registerSession(_theSession);
          }
          catch (Exception e)
          {
            Logger.Error("Error creating the session (it will be destroyed): " + e);
            if (_theSession != null)
              _theSession.destroy();
            throw;
          }
        }
        UpdateActivation();
      }
      return _theSession;
    }

    private void UpdateActivation()
    {
      if (_observer != null)
      {
        try
        {
          DeactivateObserver();
        }
        catch (Exception e)
        {
          Logger.Warn("Failed to deactivate previously activated observer: " + e);
        }
        try
        {
          OrbServices.ActivateObject((MarshalByRefObject)_observer);
          _obsId = _theSession.subscribeObserver(_observer);
          Logger.Info("Observer subscribed");
        }
        catch (Exception e)
        {
          if (_obsId != 0)
            _theSession.unsubscribeObserver(_obsId);
          Logger.Error("Error subscribing to the session: " + e);
          throw;
        }
      }
      if (_consumer != null)
      {
        try
        {
          DeactivateConsumer();
        }
        catch (Exception e)
        {
          Logger.Warn("Failed to deactivate previously activated consumer: " + e);
        }
        try
        {
          OrbServices.ActivateObject((MarshalByRefObject)_consumer);
          _subsId = _theSession.channel.subscribe(_consumer);
          Logger.Info("Consumer registered");
        }
        catch (Exception e)
        {
          if (_subsId != 0)
            _theSession.channel.unsubscribe(_subsId);
          Logger.Error("Error subscribing to the channel: " + e);
          throw;
        }
      }
    }

    private void DeactivateConsumer()
    {
      if (_subsId != 0)
      {
        _theSession.channel.unsubscribe(_subsId);
        OrbServices.DeactivateObject((MarshalByRefObject)_consumer);
        _subsId = 0;
      }
    }

    private void DeactivateObserver()
    {
      if (_obsId != 0)
      {
        _theSession.unsubscribeObserver(_obsId);
        OrbServices.DeactivateObject((MarshalByRefObject)_observer);
        _obsId = 0;
      }
    }

    public void ExitCollaboration()
    {
      lock (_locker)
      {
        try
        {
          DeactivateConsumer();
          DeactivateObserver();
          Logger.Info("Collaboration finished");
        }
        finally
        {
          _subsId = 0;
          _obsId = 0;
          _theSession = null;
          _consumer = null;
          _observer = null;
          _collabs = null;
          _sessions = null;
        }
      }
    }

    public void ShareDataKeys(List<byte[]> keys)
    {
      omg.org.CORBA.TypeCode byteTC = OrbServices.GetSingleton().create_octet_tc();
      omg.org.CORBA.TypeCode sequenceTC = OrbServices.GetSingleton().create_sequence_tc(0, byteTC);
      omg.org.CORBA.TypeCode arrayTC = OrbServices.GetSingleton().create_array_tc(keys.Count, sequenceTC);
      Share(new Any(keys.ToArray(), arrayTC));
    }

    public void Share(Any any)
    {
      try
      {
        _theSession.channel.push(any);
      }
      catch (Exception e)
      {
        Logger.Error("Error trying to push an object to the channel: " + e);
        throw;
      }
    }

    public List<byte[]> ConsumeDataKeys()
    {
      List<byte[]> list = new List<byte[]>();
      if (_consumer is Consumer)
      {
        var obj = (_consumer as Consumer);
        lock (obj.Keys)
        {
          list.AddRange(obj.Keys);
          obj.Keys.Clear();
        }
      }
      return list;
    }

    public List<object> ConsumeAnys()
    {
      List<object> list = new List<object>();
      if (_consumer is Consumer)
      {
        var obj = (_consumer as Consumer);
        lock (obj.Anys)
        {
          list.AddRange(obj.Anys);
          obj.Anys.Clear();
        }
      }
      return list;
    }

    private SessionRegistry GetSessions()
    {
      bool find = _sessions == null;
      if (!find)
      {
        try
        {
          find = !_context.ORB.non_existent(_sessions);
        }
        catch (Exception)
        {
          find = true;
        }
      }
      if (find)
      {
        ServiceProperty[] serviceProperties = new ServiceProperty[1];
        string sessionRegType = Repository.GetRepositoryID(typeof(SessionRegistry));
        serviceProperties[0] =
          new ServiceProperty("openbus.component.interface", sessionRegType);
        ServiceOfferDesc[] services = _context.OfferRegistry.findServices(serviceProperties);

        foreach (ServiceOfferDesc offerDesc in services)
        {
          try
          {
            MarshalByRefObject obj =
              offerDesc.service_ref.getFacet(sessionRegType);
            if (obj == null)
            {
              continue;
            }
            _sessions = obj as SessionRegistry;
            if (_sessions != null)
            {
              break; // found one
            }
          }
          catch (Exception e)
          {
            NO_PERMISSION npe = null;
            if (e is TargetInvocationException)
            {
              npe = e.InnerException as NO_PERMISSION;
            }
            // caso não seja uma NO_PERMISSION{NoLogin} descarta essa oferta.
            if ((npe == null) && (!(e is NO_PERMISSION)))
            {
              continue;
            }
            npe = npe ?? e as NO_PERMISSION;
            switch (npe.Minor)
            {
              case NoLoginCode.ConstVal:
                throw;
            }
          }
        }
      }
      return _sessions;
    }

    private CollaborationRegistry GetCollabs()
    {
      bool find = _collabs == null;
      if (!find)
      {
        try
        {
          find = !_context.ORB.non_existent(_collabs);
        }
        catch (Exception)
        {
          find = true;
        }
      }
      if (find)
      {
        ServiceProperty[] serviceProperties = new ServiceProperty[1];
        string collabsRegType = Repository.GetRepositoryID(typeof(CollaborationRegistry));
        serviceProperties[0] =
          new ServiceProperty("openbus.component.interface", collabsRegType);
        ServiceOfferDesc[] services = _context.OfferRegistry.findServices(serviceProperties);

        foreach (ServiceOfferDesc offerDesc in services)
        {
          try
          {
            MarshalByRefObject obj =
              offerDesc.service_ref.getFacet(collabsRegType);
            if (obj == null)
            {
              continue;
            }
            _collabs = obj as CollaborationRegistry;
            if (_collabs != null)
            {
              break; // found one
            }
          }
          catch (Exception e)
          {
            NO_PERMISSION npe = null;
            if (e is TargetInvocationException)
            {
              npe = e.InnerException as NO_PERMISSION;
            }
            // caso não seja uma NO_PERMISSION{NoLogin} descarta essa oferta.
            if ((npe == null) && (!(e is NO_PERMISSION)))
            {
              continue;
            }
            npe = npe ?? e as NO_PERMISSION;
            switch (npe.Minor)
            {
              case NoLoginCode.ConstVal:
                throw;
            }
          }
        }
      }
      return _collabs;
    }

    /// <summary>
    /// Consumidor simplificado para o canais de eventos de sessões com suporte a
    /// consumir byte[] e byte[][] como DataKeys.
    /// </summary>
    public class Consumer : MarshalByRefObject, EventConsumer
    {
      public readonly IList<byte[]> Keys;
      public readonly IList<object> Anys;

      public Consumer()
      {
        Keys = new List<byte[]>();
        Anys = new List<object>();
      }
      /// <summary>
      /// Recebe um evento encapsulado em um <c>omg.org.CORBA.Any</c> e acumula os eventos 
      /// para serem consumidos através do <see cref="EasyCollaboration.ConsumeDataKeys()"/> 
      /// ou <see cref="EasyCollaboration.ConsumeAnys()"/>.
      /// <para>Para estar disponível no <see cref="EasyCollaboration.ConsumeDataKeys()"/>,
      /// o evento precisa ser ou uma sequ~encia de bytes (byte[]) ou um array de uma sequência
      /// de bytes (byte[][]). Caso o evento seja de qualquer outro tipo estará disponível
      /// no <see cref="EasyCollaboration.ConsumeAnys()"/>.</para>
      /// </summary>
      /// <param name="e">evento enviado para o <see cref="EventChannel"/> contido na <see cref="CollaborationSession"/></param>
      public void push(object e)
      {
        Logger.Info("Received event type: " + e.GetType().ToString());
        if (e is byte[])
        {
          Keys.Add((byte[])e);
        }
        else if (e is byte[][])
        {
          foreach (byte[] item in (byte[][])e)
          {
            Keys.Add(item);
          }
        }
        else
        {
          Anys.Add(e);
        }
      }
      /// <summary>
      /// Configura a política de ciclo de vida do MarshalByRefObject.
      /// </summary>
      /// <returns>sempre retorna null</returns>
      public override object InitializeLifetimeService()
      {
        // Evita a desativação automática pela política de ciclo de vida do MarshalByRefObject
        return null;
      }
    }

    /// <summary>
    /// Observador simplificado para sessões de colaboração.
    /// </summary>
    public class Observer : MarshalByRefObject, CollaborationObserver
    {

      public void memberAdded(string name, IComponent member)
      {
        Logger.Info("Member added: " + name);
      }

      public void memberRemoved(string name)
      {
        Logger.Info("Member removed: " + name);
      }

      public void destroyed()
      {
        Logger.Info("Session destroyed");
      }

      /// <summary>
      /// Configura a política de ciclo de vida do MarshalByRefObject.
      /// </summary>
      /// <returns>sempre retorna null</returns>
      public override object InitializeLifetimeService()
      {
        // Evita a desativação automática pela política de ciclo de vida do MarshalByRefObject
        return null;
      }
    }
  }
}

File: java/src/main/java/lib/EasyCollaboration.java
package lib;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.logging.Logger;

import org.omg.CORBA.Any;
import org.omg.CORBA.ORB;
import org.omg.CORBA.OctetSeqHelper;
import org.omg.CORBA.ORBPackage.InvalidName;
import org.omg.PortableServer.POA;
import org.omg.PortableServer.POAHelper;

import scs.core.IComponent;
import tecgraf.openbus.OpenBusContext;
import tecgraf.openbus.core.v2_0.services.ServiceFailure;
import tecgraf.openbus.core.v2_0.services.offer_registry.ServiceOfferDesc;
import tecgraf.openbus.core.v2_0.services.offer_registry.ServiceProperty;
import tecgraf.openbus.services.collaboration.v1_0.CollaborationObserver;
import tecgraf.openbus.services.collaboration.v1_0.CollaborationObserverHelper;
import tecgraf.openbus.services.collaboration.v1_0.CollaborationObserverPOA;
import tecgraf.openbus.services.collaboration.v1_0.CollaborationRegistry;
import tecgraf.openbus.services.collaboration.v1_0.CollaborationRegistryHelper;
import tecgraf.openbus.services.collaboration.v1_0.CollaborationSession;
import tecgraf.openbus.services.collaboration.v1_0.EventConsumer;
import tecgraf.openbus.services.collaboration.v1_0.EventConsumerHelper;
import tecgraf.openbus.services.collaboration.v1_0.EventConsumerPOA;
import tecgraf.openbus.services.collaboration.v1_0.SessionDoesNotExist;
import tecgraf.openbus.services.collaboration.v1_0.SessionRegistry;
import tecgraf.openbus.services.collaboration.v1_0.SessionRegistryHelper;

public class EasyCollaboration implements IEasyCollaboration {

  private OpenBusContext context;
  private SessionRegistry sessions;
  private CollaborationRegistry collabs;
  private CollaborationSession theSession;
  private Consumer servant;
  private SessionObserver observer;
  private EventConsumer consumer;
  private int subsId;
  private int obsId;
  
  /**
   * O <i>logger</i>.
   */
  private static Logger logger = Logger
    .getLogger(EasyCollaboration.class.getName());

  public EasyCollaboration(OpenBusContext context) {
    this.context = context;
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public CollaborationSession startCollaboration() throws ServiceFailure {
    logger.info("Starting collaboration");
    SessionRegistry sreg = getSessions();
    try {
      theSession = sreg.getSession();
      logger.info("Session retrieved: " + theSession);
    }
    catch (SessionDoesNotExist e) {
      logger.warning("Session not found for entity " + e.entity);
    }
    catch (tecgraf.openbus.core.v2_0.services.ServiceFailure e) {
      throw new ServiceFailure(e.getMessage());
    }
    catch (Throwable t) {
      logger.severe("Unknown error: " + t.getMessage());
      t.printStackTrace();
    }
    try {
      if (theSession == null) {
        CollaborationRegistry collab = getCollabs();
        theSession = collab.createCollaborationSession();
        sreg.registerSession(theSession);
      }

      obsId = theSession.subscribeObserver(buildObserver());
      logger.info("Observer subscribed");

      consumer = buildConsumer();
      subsId = theSession.channel().subscribe(consumer);
      logger.info("Consumer registered");
    }
    catch (tecgraf.openbus.core.v2_0.services.ServiceFailure ex) {
      throw new ServiceFailure(ex.getMessage());
    } 
    return theSession;

  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void exitCollaboration() throws ServiceFailure {
    try {
      theSession.channel().unsubscribe(subsId);
      theSession.unsubscribeObserver(obsId);
      logger.info("Collaboration finished");
    }
    catch (tecgraf.openbus.core.v2_0.services.ServiceFailure e) {
      throw new ServiceFailure(e.getMessage());
    }
    finally {
      
      subsId = 0;
      obsId = 0;
      theSession = null;
      servant = null;
      consumer = null;
      observer = null;
      collabs = null;
      sessions = null;
    }
  }

  private CollaborationObserver buildObserver() throws ServiceFailure {
    try {
      ORB orb = context.orb();
      observer = new SessionObserver(context.getCurrentConnection().login().entity);
      POA poa = POAHelper.narrow(orb.resolve_initial_references("RootPOA"));
      poa.the_POAManager().activate();
      byte[] id = poa.activate_object(observer);
      CollaborationObserver ref = CollaborationObserverHelper.narrow(poa.id_to_reference(id));
      return ref;      
    }
    catch (Exception e) {
      // should never happen
      throw new ServiceFailure(e.getLocalizedMessage(),
          "Error while session observer activation");
    }
  }
  
  private EventConsumer buildConsumer() throws ServiceFailure {
    try {
      ORB orb = context.orb();
      servant = new Consumer();
      POA poa = POAHelper.narrow(orb.resolve_initial_references("RootPOA"));
      poa.the_POAManager().activate();
      byte[] id = poa.activate_object(servant);
      EventConsumer ref = EventConsumerHelper.narrow(poa.id_to_reference(id));
      return ref;
    }
    catch (Exception e) {
      // should never happen
      throw new ServiceFailure(e.getLocalizedMessage(),
        "Error while event consumer activation");
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void shareDataKey(byte[] key) throws ServiceFailure {
    try {
      Any any = context.orb().create_any();
      OctetSeqHelper.insert(any, key);
      theSession.channel().push(any);
    }
    catch (tecgraf.openbus.core.v2_0.services.ServiceFailure e) {
      throw new ServiceFailure(e.getMessage());
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public void shareDataKeys(List<byte[]> keys) throws ServiceFailure {
    try {
      for(byte[] key: keys) {
        Any any = context.orb().create_any();
        OctetSeqHelper.insert(any, key);
        theSession.channel().push(any);
      }
    }
    catch (tecgraf.openbus.core.v2_0.services.ServiceFailure e) {
      throw new ServiceFailure(e.getMessage());
    }
  }
  
  /**
   * {@inheritDoc}
   */
  @Override
  public void shareAny(Any any) throws ServiceFailure {
    try {
      theSession.channel().push(any);
    }
    catch (tecgraf.openbus.core.v2_0.services.ServiceFailure e) {
      throw new ServiceFailure(e.getMessage());
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public List<byte[]> consumeDataKeys() {
    synchronized (servant.keys) {
      LinkedList<byte[]> list = new LinkedList<byte[]>(servant.keys);
      servant.keys.clear();
      return list;
    }
  }

  /**
   * {@inheritDoc}
   */
  @Override
  public List<Any> consumeAnys() {
    synchronized (servant.anys) {
      LinkedList<Any> list = new LinkedList<Any>(servant.anys);
      servant.anys.clear();
      return list;
    }
  }

  private SessionRegistry getSessions() throws ServiceFailure {
    if (sessions == null || sessions._non_existent()) {
      ServiceProperty[] serviceProperties = new ServiceProperty[1];
      serviceProperties[0] =
        new ServiceProperty("openbus.component.interface",
          SessionRegistryHelper.id());
      List<ServiceOfferDesc> services =
        LibUtils.findOffer(context.getOfferRegistry(), serviceProperties, 1,
          10, 1);

      for (ServiceOfferDesc offerDesc : services) {
        org.omg.CORBA.Object obj =
          offerDesc.service_ref.getFacet(SessionRegistryHelper.id());
        if (obj == null) {
          continue;
        }
        sessions = SessionRegistryHelper.narrow(obj);
        if (sessions != null) {
          break; // found one
        }
      }
    }
    return sessions;
  }

  private CollaborationRegistry getCollabs() throws ServiceFailure {
    if (collabs == null || collabs._non_existent()) {
      ServiceProperty[] serviceProperties = new ServiceProperty[1];
      serviceProperties[0] =
        new ServiceProperty("openbus.component.interface",
          CollaborationRegistryHelper.id());
      List<ServiceOfferDesc> services =
        LibUtils.findOffer(context.getOfferRegistry(), serviceProperties, 1,
          10, 1);

      for (ServiceOfferDesc offerDesc : services) {
        org.omg.CORBA.Object obj =
          offerDesc.service_ref.getFacet(CollaborationRegistryHelper.id());
        if (obj == null) {
          continue;
        }
        collabs = CollaborationRegistryHelper.narrow(obj);
        if (collabs != null) {
          break; // found one
        }
      }
    }
    return collabs;
  }

  /**
   * Consumidor simplificado para o canais de eventos de sessões
   * 
   * @author Tecgraf/PUC-Rio
   *
   */
  class Consumer extends EventConsumerPOA {

    private List<byte[]> keys;
    private List<Any> anys;

    public Consumer() {
      this.keys = Collections.synchronizedList(new LinkedList<byte[]>());
      this.anys = Collections.synchronizedList(new LinkedList<Any>());
    }

    @Override
    public void push(Any event)
      throws tecgraf.openbus.core.v2_0.services.ServiceFailure {
      logger.info("Received event");
      if (event.type().equal(OctetSeqHelper.type())) {
        keys.add(OctetSeqHelper.extract(event));
      }
      else {
        anys.add(event);
      }
    }
  }
  
  /**
   * Observador simplificado para sessões de colaboração.
   * 
   * 
   * @author Tecgraf/PUC-Rio
   *
   */
  class SessionObserver extends CollaborationObserverPOA {

    private String entity;

    public SessionObserver(String entity) {
      this.entity = entity;
    }

    @Override
    public void memberAdded(String name, IComponent member) throws tecgraf.openbus.core.v2_0.services.ServiceFailure{
      logger.info("Member added: " + name);
    }

    @Override
    public void memberRemoved(String name) throws tecgraf.openbus.core.v2_0.services.ServiceFailure {
      logger.info("Member removed: " + name);
    }

    @Override
    public void destroyed() throws tecgraf.openbus.core.v2_0.services.ServiceFailure {
      logger.info("Session destroyed");
    }
  }

}
